data-science-onramp/data-ingestion/setup.py

```diff
@@ -0,0 +1,202 @@
+    return random.choices([seconds, minutes, hours,
+                          str(random.randint(-1000, -1))],
+                          weights=[0.3, 0.3, 0.3, 0.1])[0]
+
+
+def station_name(name):
+    '''Replaces '&' with '/' with a 50% chance'''
+    if not name:
+        return None
+    return random.choice([name, name.replace("&", "/")])
+
+
+def user_type(user):
+    '''Manipulates the user type string'''
+    if not user:
+        return None
+    return random.choice([user, user.upper(), user.lower(),
+                          "sub" if user == "Subscriber" else user,
+                          "cust" if user == "Customer" else user])
+
+
+def gender(s):
+    '''Manipulates the gender string'''
+    if not s:
+        return None
+    return random.choice([s.upper(), s.lower(),
+                         s[0].upper() if len(s) > 0 else "",
+                         s[0].lower() if len(s) > 0 else ""])
+
+
+def convert_angle(angle):
+    '''Converts long and lat to DMS notation'''
+    if angle is None:
+        return None
+    degrees = int(angle)
+    minutes = int((angle - degrees) * 60)
+    seconds = int((angle - degrees - minutes/60) * 3600)
+    new_angle = str(degrees) + "\u00B0" + \
+        str(minutes) + "'" + str(seconds) + '"'
+    return random.choices([str(angle), new_angle],
+                          weights=[0.55, 0.45])[0]
+
+
+def create_bigquery_dataset():
+    # Create BigQuery Dataset
+    client = bigquery.Client()
+    dataset_id = f'{client.project}.{DATASET_NAME}'
+    dataset = bigquery.Dataset(dataset_id)
+    dataset.location = "US"
+    dataset = client.create_dataset(dataset)
+
+
+def write_to_bigquery(df, table_name):
+    '''Write a dataframe to BigQuery'''
+    client = bigquery.Client()
+    dataset_id = f'{client.project}.{DATASET_NAME}'
+
+    # Saving the data to BigQuery
+    df.write.format('bigquery') \
+        .option('table', f"{dataset_id}.{table_name}") \
+        .save()
+
+    print(f"Table {table_name} successfully written to BigQuery")
+
+
+def main():
+    # Create a SparkSession under the name "setup". Viewable via the Spark UI
+    spark = SparkSession.builder.appName("setup").getOrCreate()
+
+    spark.conf.set('temporaryGcsBucket', BUCKET_NAME)
+
+    create_bigquery_dataset()
+
+    # Whether we are running the job as a test
+    test = False
+
+    # Check whether or not the job is running as a test
+    if '--test' in sys.argv:
+        test = True
+        print("A subset of the whole dataset will be uploaded to BigQuery")
+    else:
+        print("Results will be uploaded to BigQuery")
+
+    # Ingest External Datasets
+    for table_name, data in EXTERNAL_TABLES.items():
+        df = spark.createDataFrame(pd.read_csv(data["url"]),
+                                   schema=data["schema"])
+
+        write_to_bigquery(df, table_name)
+
+    # Check if table exists
+    try:
+        df = spark.read.format('bigquery').option('table', TABLE).load()
+        if test:
```

Comment:
nit - add a comment that lets the user know what happens and why this test flag is here